python/meterpreter/ext_server_stdapi.py

```diff
@@ -1712,6 +1732,54 @@ def stdapi_fs_mount_show(request, response):
         response += tlv_pack(TLV_TYPE_MOUNT_GROUP, mount)
     return ERROR_SUCCESS, response
 
+@register_function_if(has_windll)
+def stdapi_net_config_get_arp_table(request, response):
+    iphlpapi = None
+    try:
+        iphlpapi = ctypes.WinDLL('iphlpapi.dll')
+    except:
+        return ERROR_FAILURE, response
+
+    ERROR_INSUFFICIENT_BUFFER = 0x0000007a
+    ERROR_NOT_SUPPORTED = 0x00000032
+    ERROR_NO_DATA = 0x000000e8
+    NO_ERROR = ERROR_SUCCESS
+    
+    MIB_IPNET_TYPE_DYNAMIC = 3
+    MIB_IPNET_TYPE_STATIC  = 4
+
+    GetIpNetTable = iphlpapi['GetIpNetTable']
+    GetIpNetTable.argtypes = [ctypes.c_void_p, ctypes.POINTER(ctypes.c_ulong), ctypes.c_long]
+    GetIpNetTable.restype = ctypes.c_ulong
+
+    ipnet_table = None
+    size = ctypes.c_ulong(0)
+    result = GetIpNetTable(ipnet_table, size, False)
+
+    if result == ERROR_INSUFFICIENT_BUFFER:
+        ipnet_table = ctypes.cast(ctypes.create_string_buffer(b'', size.value), ctypes.POINTER(TABLE_IPNETTABLE(size.value)))
```

Comment:
This isn't quite right. The second parameter `SizePointer` from [`GetIpNetTable`](https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getipnettable#return-value) is the size in bytes, not the number of elements. This means when you pass it to `TABLE_IPNETTABLE` you're allocating quite a bit more memory than you need. I'm not 100% sure you even need to cast it to the `MIB_IPNETTABLE` at this point anyways. You may be able to use cast it as a void pointer.

If you can figure out to to void casting this to `MIB_IPNETTABLE` here, then you could remove the `TABLE_IPNETTABLE` function which is a bit awkward because it's not actually a datatype and instead inline the definition of `MIB_IPNETTABLE` once you do know how many elements are necessary. That would be my suggestion so the correct amount of space is allocated and you don't need to calculate the number of elements manually.

Reply:
You right! `TABLE_IPNETTABLE(size.value)` in line 1760 always allocates more than enough. Here is the fix:
```
if result == ERROR_INSUFFICIENT_BUFFER:
    ipnet_table = ctypes.cast(ctypes.create_string_buffer(b'', size.value), ctypes.c_void_p)

...

dwNumEntries = ctypes.cast(ipnet_table.value, ctypes.POINTER(ctypes.c_ulong)).contents.value
ipnet_table = ctypes.cast(ipnet_table, ctypes.POINTER(TABLE_IPNETTABLE(dwNumEntries)))
```
Thanks again <3